enh: Adds support for `Time` datatype

marvinl803 opened the pull request:
<!--
# Thanks for contributing a pull request! 
## Please make sure you see our contribution guidelines: https://github.com/narwhals-dev/narwhals/blob/main/CONTRIBUTING.md
-->

## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [X] 🐳 Other

## Related issues

- Related issue #1989

## Checklist

- [X] Code follows style guide (ruff)
- [X] Tests added
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below
I have added support for the Polars Time datatype. I wanted to implement this first to ensure there are no issues before proceeding with the Binary datatype, as mentioned in the issue.

Please let me know if any changes or fixes are needed. Thank you!

FBruzzesi commented on narwhals/_polars/utils.py:
Can we add the conversion from narwhals `Time` to polars `Time` in `narwhals_to_native_dtype` function?

marvinl803 replied:
Sounds good! Thank you so much for your feedback. I'll start implementing your suggestions!

FBruzzesi reviewed (commented):
Thanks for the contribution @marvinl803 - This is a good start 🙏🏼

I would like to aim to support the following in this PR:

- Add a case test in [`tests/expr_and_series/cast_test.py`](https://github.com/narwhals-dev/narwhals/blob/main/tests/expr_and_series/cast_test.py)
- Support more than just polars, as specified in [#1989 comment](https://github.com/narwhals-dev/narwhals/issues/1989#issuecomment-2658792772), pyarrow and duckdb should be possible to support
- Make sure that conversion from and to the backend is supported

marvinl803 commented:
> Thanks for the contribution @marvinl803 - This is a good start 🙏🏼
> 
> I would like to aim to support the following in this PR:
> 
> * Add a case test in [`tests/expr_and_series/cast_test.py`](https://github.com/narwhals-dev/narwhals/blob/main/tests/expr_and_series/cast_test.py)
> * Support more than just polars, as specified in [#1989 comment](https://github.com/narwhals-dev/narwhals/issues/1989#issuecomment-2658792772), pyarrow and duckdb should be possible to support
> * Make sure that conversion from and to the backend is supported

Hi @FBruzzesi,

I have added the test cases where you suggested and also implemented support for PyArrow and DuckDB. However, I encountered an issue in narwhals/dtypes.py—I wasn’t sure how to provide an example for the docs using DuckDB.

Additionally, I added Time to all the other supported types in narwhals_to_native_dtype, except for PyArrow. I’m uncertain whether we should return time32 or time64 in that case. Could you provide some guidance on how best to handle this?

I appreciate your help!

FBruzzesi commented:
> I have added the test cases where you suggested and also implemented support for PyArrow and DuckDB. However, I encountered an issue in narwhals/dtypes.py—I wasn’t sure how to provide an example for the docs using DuckDB.

Thanks for your adjustments - I think we are very close to have this ready. Regarding the narwhals failing tests:

- One is because of coverage: conversion from narwhals to native has no test yet
- The other one if because of top level imports - try use `constructor` and mark unsupported backends as `xfail` rather than manually instantiating dataframes

> Additionally, I added Time to all the other supported types in narwhals_to_native_dtype, except for PyArrow. I’m uncertain whether we should return time32 or time64 in that case. Could you provide some guidance on how best to handle this?

I think we should convert narwhals Time to pyarrow time64 since that's what [polars Time](https://docs.pola.rs/api/python/stable/reference/api/polars.datatypes.Time.html#polars.datatypes.Time) mentions:

> _The underlying representation of this type is a 64-bit signed integer. The integer indicates the number of nanoseconds since midnight._

--- 

As a side note, we will need to make PRs in shiny and marimo to fix their failing tests

FBruzzesi commented on narwhals/_pandas_like/utils.py:
```diff
@@ -412,6 +412,8 @@ def non_object_native_to_narwhals_dtype(dtype: str, version: Version) -> DType:
         return dtypes.Date()
     if dtype.startswith("decimal") and dtype.endswith("[pyarrow]"):
         return dtypes.Decimal()
+    if dtype.startswith("time") and dtype.endswith("[pyarrow]"):
+        return dtypes.Time()
```
Pattern is `time<32|64>[<unit>][pyarrow]`

Example:
```py
from datetime import time
import pandas as pd
import pyarrow as pa

data = {"a": [time(12, 0, 0), time(12, 0, 5)]}

pd.DataFrame(data).convert_dtypes(dtype_backend="pyarrow").astype(pd.ArrowDtype(pa.time64("ns"))).dtypes
a    time64[ns][pyarrow]
dtype: object
```

FBruzzesi commented on narwhals/_pandas_like/utils.py:
```diff
@@ -610,7 +612,9 @@ def narwhals_to_native_dtype(  # noqa: PLR0915
+    if isinstance_or_issubclass(
+        dtype, (dtypes.Struct, dtypes.Array, dtypes.List, dtypes.Time)
+    ):
         if implementation is Implementation.PANDAS and backend_version >= (2, 2):
```
Unrelated, but... should we check modin backed by pyarrow?

FBruzzesi commented:
@marvinl803 I added a few commits to reach coverage and add pandas support (if backed by pyarrow dtype).

I think this is ready to merge, but we should have ready the PRs to address marimo and shiny whenever we press the merge button. I will try to find a bit of time tomorrow (no promises 🙈)

dangotbanned commented on tests/expr_and_series/cast_test.py:
```diff
@@ -285,3 +286,16 @@ def test_raise_if_polars_dtype(constructor: Constructor, dtype: Any) -> None:
         df.select(nw.col("a").cast(dtype))
+
+
+def test_cast_time(request: pytest.FixtureRequest, constructor: Constructor) -> None:
+    if "pandas" in str(constructor) and PANDAS_VERSION < (2, 2):
+        request.applymarker(pytest.mark.xfail)
+
+    if any(backend in str(constructor) for backend in ("dask", "pyspark", "modin")):
+        request.applymarker(pytest.mark.xfail)
+
+    data = {"a": [time(12, 0, 0), time(12, 0, 5)]}
+    df = nw.from_native(constructor(data))
+    result = df.select(nw.col("a").cast(nw.Time()))
+    assert result.collect_schema() == {"a": nw.Time()}
```
Great to see this is working for the `Expr.cast` case!

Could we also gets some tests to see how `nw.Time` interacts with [`Expr.dt`](https://narwhals-dev.github.io/narwhals/api-reference/expr_dt/) methods?
I'm thinking we should aim to support anything that **doesn't** mention *date*, *duration*, *time_zone* components

dangotbanned replied:
The [`polars` Temporal](https://docs.pola.rs/api/python/stable/reference/expressions/temporal.html) docs usually specify which subset of `TemporalType` they work with

dangotbanned replied:
I've only found [`polars.Expr.dt.to_string`](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.dt.to_string.html) mention `Time` so far

dangotbanned replied:
There's also a huge suite here to draw from for test ideas: https://github.com/pola-rs/polars/blob/52b93ef5909cd0a5790001894aec4b473c361631/py-polars/tests/unit/datatypes/test_temporal.py

FBruzzesi replied:
That might get more complex that expected: pandas does not access pyarrow time via `.dt`:

```py
 from datetime import time
import pandas as pd
import pyarrow as pa

data = {"a": [time(12, 0, 0), time(12, 0, 5)]}

(pd.DataFrame(data)
.convert_dtypes(dtype_backend="pyarrow")
.astype(pd.ArrowDtype(pa.time64("ns")))
["a"].dt
)
```

> AttributeError: Can only use .dt accessor with datetimelike values

dangotbanned replied:
@FBruzzesi we don't *have* to resolve all the issues in this PR - but I think it would still be worth identifying them for follow-ups.

Before reading through (https://github.com/narwhals-dev/narwhals/pull/2113#discussion_r1984715599) I wasn't aware of how little support there was for `Time`

marvinl803 commented:
> @marvinl803 I added a few commits to reach coverage and add pandas support (if backed by pyarrow dtype).
> 
> I think this is ready to merge, but we should have ready the PRs to address marimo and shiny whenever we press the merge button. I will try to find a bit of time tomorrow (no promises 🙈)
> 
> Plotly failure is unrelated, they simply reworked some deps files

Got it! Thanks for your help and the updates. I'm not too familiar with Marimo and Shiny, but happy to assist however I can. Let me know if there's anything I can do! 👍

FBruzzesi commented:
> Got it! Thanks for your help and the updates, @FBruzzesi. I'm not too familiar with Marimo and Shiny, but happy to assist however I can. Let me know if there's anything I can do! 👍

Hey @marvinl803 I opened the one PR in each repo, and forgot to mention it. You should be able to see the link somewhere here (I am from mobile and it's tricky to find the links) 😊

marvinl803 commented:
> Hey @marvinl803 I opened one PR in each repo, and forgot to mention it. You should be able to see the link somewhere here (I am from mobile and it's tricky to find the links) 😊

No problem, @FBruzzesi! I found them—I'll check them out.

FBruzzesi reviewed (approved):
Thanks a ton for the contribution (and your patience) @marvinl803 

Let's ship this to include this feature in today's release
![WooblesTheWooblesGIF](https://github.com/user-attachments/assets/34e55d41-f331-4c43-9ea0-3b167f50c17b)


P.S. @dangotbanned I was not able to solve the mypy issues is CI 🥲

dangotbanned commented:
> P.S. @dangotbanned I was not able to solve the mypy issues is CI 🥲

@FBruzzesi Let me take a quick look before merging 🤝

dangotbanned commented:
I've opened a PR upstream for another fix

- https://github.com/zen-xu/pyarrow-stubs/pull/190

FBruzzesi commented:
Thanks @dangotbanned , shall we merge?
